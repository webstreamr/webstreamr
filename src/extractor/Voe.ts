import bytes from 'bytes';
import * as cheerio from 'cheerio';
import { NotFoundError } from '../error';
import { Context, Format, Meta, UrlResult } from '../types';
import {
  buildMediaFlowProxyExtractorStreamUrl, guessHeightFromPlaylist,
  supportsMediaFlowProxy,
} from '../utils';
import { Extractor } from './Extractor';

/** @see https://github.com/Gujal00/ResolveURL/blob/master/script.module.resolveurl/lib/resolveurl/plugins/voesx.py */
export class Voe extends Extractor {
  public readonly id = 'voe';

  public readonly label = 'VOE (via MediaFlow Proxy)';

  public override viaMediaFlowProxy = true;

  public supports(ctx: Context, url: URL): boolean {
    const supportedDomain = null !== url.host.match(/voe/)
      || [
        '19turanosephantasia.com',
        '20demidistance9elongations.com',
        '30sensualizeexpression.com',
        '321naturelikefurfuroid.com',
        '35volitantplimsoles5.com',
        '449unceremoniousnasoseptal.com',
        '745mingiestblissfully.com',
        'adrianmissionminute.com',
        'alleneconomicmatter.com',
        'antecoxalbobbing1010.com',
        'apinchcaseation.com',
        'audaciousdefaulthouse.com',
        'availedsmallest.com',
        'bigclatterhomesguideservice.com',
        'boonlessbestselling244.com',
        'bradleyviewdoctor.com',
        'brittneystandardwestern.com',
        'brucevotewithin.com',
        'chromotypic.com',
        'chuckle-tube.com',
        'cindyeyefinal.com',
        'counterclockwisejacky.com',
        'crownmakermacaronicism.com',
        'cyamidpulverulence530.com',
        'diananatureforeign.com',
        'donaldlineelse.com',
        'edwardarriveoften.com',
        'erikcoldperson.com',
        'figeterpiazine.com',
        'fittingcentermondaysunday.com',
        'fraudclatterflyingcar.com',
        'gamoneinterrupted.com',
        'generatesnitrosate.com',
        'goofy-banana.com',
        'graceaddresscommunity.com',
        'greaseball6eventual20.com',
        'guidon40hyporadius9.com',
        'heatherdiscussionwhen.com',
        'housecardsummerbutton.com',
        'jamessoundcost.com',
        'jamiesamewalk.com',
        'jasminetesttry.com',
        'jayservicestuff.com',
        'jennifercertaindevelopment.com',
        'jilliandescribecompany.com',
        'johnalwayssame.com',
        'jonathansociallike.com',
        'josephseveralconcern.com',
        'kathleenmemberhistory.com',
        'kellywhatcould.com',
        'kennethofficialitem.com',
        'kristiesoundsimply.com',
        'launchreliantcleaverriver.com',
        'lisatrialidea.com',
        'loriwithinfamily.com',
        'lukecomparetwo.com',
        'lukesitturn.com',
        'mariatheserepublican.com',
        'matriculant401merited.com',
        'maxfinishseveral.com',
        'metagnathtuggers.com',
        'michaelapplysome.com',
        'mikaylaarealike.com',
        'nathanfromsubject.com',
        'nectareousoverelate.com',
        'nonesnanking.com',
        'paulkitchendark.com',
        'realfinanceblogcenter.com',
        'rebeccaneverbase.com',
        'reputationsheriffkennethsand.com',
        'richardsignfish.com',
        'roberteachfinal.com',
        'robertordercharacter.com',
        'robertplacespace.com',
        'sandratableother.com',
        'sandrataxeight.com',
        'scatch176duplicities.com',
        'sethniceletter.com',
        'shannonpersonalcost.com',
        'simpulumlamerop.com',
        'stevenimaginelittle.com',
        'strawberriesporail.com',
        'telyn610zoanthropy.com',
        'timberwoodanotia.com',
        'toddpartneranimal.com',
        'toxitabellaeatrebates306.com',
        'uptodatefinishconferenceroom.com',
        'v-o-e-unblock.com',
        'valeronevijao.com',
        'wolfdyslectic.com',
        'yodelswartlike.com',
      ].includes(url.host);

    return supportedDomain && supportsMediaFlowProxy(ctx);
  }

  public override normalize(url: URL): URL {
    return new URL(`/${url.pathname.split('/').at(-1)}`, url);
  }

  protected async extractInternal(ctx: Context, url: URL, meta: Meta): Promise<UrlResult[]> {
    const headers = { Referer: meta.referer ?? url.href };

    let html: string;
    try {
      html = await this.fetcher.text(ctx, url, { headers });
    } catch (error) {
      /* istanbul ignore next */
      if (error instanceof NotFoundError && !url.href.includes('/e/')) {
        return await this.extractInternal(ctx, new URL(`/e${url.pathname}`, url.origin), meta);
      }

      /* istanbul ignore next */
      throw error;
    }

    const redirectMatch = html.match(/window\.location\.href\s*=\s*'([^']+)/);
    if (redirectMatch && redirectMatch[1]) {
      return await this.extractInternal(ctx, new URL(redirectMatch[1]), meta);
    }

    if (/An error occurred during encoding/.test(html)) {
      throw new NotFoundError();
    }

    const $ = cheerio.load(html);
    const title = $('meta[name="description"]').attr('content')?.trim().replace(/^Watch /, '').replace(/ at VOE$/, '').trim();

    const sizeMatch = html.matchAll(/[\d.]+ ?[GM]B/g).toArray().at(-1);
    const size = sizeMatch ? bytes.parse(sizeMatch[0] as string) as number : /* istanbul ignore next */null;

    const playlistUrl = await buildMediaFlowProxyExtractorStreamUrl(ctx, this.fetcher, 'Voe', url, headers);

    const heightMatch = html.match(/<b>(\d{3,})p<\/b>/);
    const height = heightMatch ? parseInt(heightMatch[1] as string) : await guessHeightFromPlaylist(ctx, this.fetcher, playlistUrl, url);

    return [
      {
        url: playlistUrl,
        format: Format.hls,
        label: this.label,
        sourceId: `${this.id}_${meta.countryCodes?.join('_')}`,
        ttl: this.ttl,
        meta: {
          ...meta,
          height,
          title,
          ...(size && size > 16777216 && { bytes: size }),
        },
      },
    ];
  };
}
