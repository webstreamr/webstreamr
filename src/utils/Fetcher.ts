import { gunzipSync, gzipSync } from 'zlib';
import TTLCache from '@isaacs/ttlcache';
import { Mutex, Semaphore, SemaphoreInterface, withTimeout } from 'async-mutex';
import CachePolicy from 'http-cache-semantics';
import { LRUCache } from 'lru-cache';
import { Cookie, CookieJar } from 'tough-cookie';
import { fetch, Headers, RequestInit, Response } from 'undici';
import winston from 'winston';
import { BlockedError, HttpError, NotFoundError, QueueIsFullError, TimeoutError, TooManyRequestsError, TooManyTimeoutsError } from '../error';
import { BlockedReason, Context } from '../types';
import { envGet } from './env';

interface HttpCacheItem {
  policy: CachePolicy;
  status: number;
  statusText: string;
  body: string;
}

interface FlareSolverrResult {
  status: string;
  message: string;
  solution: {
    url: string;
    status: number;
    cookies: {
      domain: string;
      expires: number;
      httpOnly: boolean;
      name: string;
      path: string;
      sameSite: string;
      secure: boolean;
      value: string;
    }[];
    userAgent: string;
    headers: Record<string, string>;
    response: string;
  };
  startTimeStamp: number;
  endTimeStamp: number;
  version: string;
}

export type CustomRequestInit = RequestInit & {
  timeoutsCountThrow?: number;
  noFlareSolverr?: boolean;
  queueLimit?: number;
  queueTimeout?: number;
  timeout?: number;
};

export class Fetcher {
  private readonly MIN_CACHE_TTL = 900000; // 15m
  private readonly DEFAULT_TIMEOUT = 10000;
  private readonly DEFAULT_QUEUE_LIMIT = 5;
  private readonly DEFAULT_QUEUE_TIMEOUT = 5000;
  private readonly DEFAULT_FAILED_REQUEST_COUNT_THROW = 10;
  private readonly TIMEOUT_CACHE_TTL = 3600000; // 1h

  private readonly logger: winston.Logger;

  private readonly httpCache = new LRUCache<string, Buffer>({ max: 1024 });
  private readonly rateLimitedCache = new TTLCache<string, undefined>({ max: 1024 });
  private readonly semaphores = new Map<string, SemaphoreInterface>();
  private readonly hostUserAgentMap = new Map<string, string>();
  private readonly cookieJar = new CookieJar();

  private readonly timeoutsCountCache = new TTLCache<string, number>({ max: 1024 });
  private readonly timeoutsCountMutex = new Mutex();

  public constructor(logger: winston.Logger) {
    this.logger = logger;
  }

  public async text(ctx: Context, url: URL, init?: CustomRequestInit): Promise<string> {
    return (await this.cachedFetch(ctx, url, init)).body;
  };

  public async textPost(ctx: Context, url: URL, body: string, init?: CustomRequestInit): Promise<string> {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'POST', body })).body;
  };

  public async head(ctx: Context, url: URL, init?: CustomRequestInit): Promise<CachePolicy.Headers> {
    return (await this.cachedFetch(ctx, url, { ...init, method: 'HEAD' })).policy.responseHeaders();
  };

  private getInit(ctx: Context, url: URL, init?: CustomRequestInit): CustomRequestInit {
    const cookieString = this.cookieJar.getCookieStringSync(url.href);

    return {
      ...init,
      headers: {
        'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
        'Accept-Language': 'en',
        'Priority': 'u=0',
        'User-Agent': this.hostUserAgentMap.get(url.host) ?? 'node',
        ...(cookieString && { Cookie: cookieString }),
        ...(ctx.ip && {
          'Forwarded': `for=${ctx.ip}`,
          'X-Forwarded-For': ctx.ip,
          'X-Forwarded-Proto': url.protocol.slice(0, -1),
          'X-Real-IP': ctx.ip,
        }),
        ...init?.headers,
      },
    };
  };

  private async handleHttpCacheItem(ctx: Context, httpCacheItem: HttpCacheItem, url: URL, init?: CustomRequestInit): Promise<HttpCacheItem> {
    if (httpCacheItem.status && httpCacheItem.status >= 200 && httpCacheItem.status <= 399) {
      return httpCacheItem;
    }

    if (httpCacheItem.status === 404) {
      throw new NotFoundError();
    }

    const responseHeaders = httpCacheItem.policy.responseHeaders();

    if (httpCacheItem.policy.responseHeaders()['cf-mitigated'] === 'challenge') {
      const noFlareSolverr = init?.noFlareSolverr ?? false;
      const flareSolverrEndpoint = envGet('FLARESOLVERR_ENDPOINT');
      if (noFlareSolverr || !flareSolverrEndpoint) {
        throw new BlockedError(BlockedReason.cloudflare_challenge, httpCacheItem.policy.responseHeaders());
      }

      this.logger.info(`Query FlareSolverr for ${url.href}`, ctx);

      const body = { cmd: 'request.get', url: url.href, session: 'default' };
      const challengeResult = await (await this.queuedFetch(ctx, new URL(flareSolverrEndpoint), { method: 'POST', body: JSON.stringify(body), headers: { 'Content-Type': 'application/json' }, queueLimit: 1, timeout: 15000 })).json() as FlareSolverrResult;
      if (challengeResult.status !== 'ok') {
        this.logger.warn(`FlareSolverr issue: ${JSON.stringify(challengeResult)}`, ctx);
        throw new BlockedError(BlockedReason.flaresolverr_failed, {});
      }

      challengeResult.solution.cookies.forEach((cookie) => {
        if (!['cf_clearance'].includes(cookie.name)) {
          return;
        }

        this.cookieJar.setCookie(
          new Cookie({
            key: cookie.name,
            value: cookie.value,
            expires: new Date(cookie.expires * 1000),
            domain: cookie.domain.replace(/^.+/, ''),
          }),
          url.href,
        );
      });

      this.hostUserAgentMap.set(url.host, challengeResult.solution.userAgent);

      httpCacheItem.body = challengeResult.solution.response;

      this.cacheSet(this.determineCacheKey(url, init), httpCacheItem);

      return httpCacheItem;
    }

    if (httpCacheItem.status === 403) {
      if (ctx.config.mediaFlowProxyUrl && url.href.startsWith(ctx.config.mediaFlowProxyUrl)) {
        throw new BlockedError(BlockedReason.media_flow_proxy_auth, httpCacheItem.policy.responseHeaders());
      }

      throw new BlockedError(BlockedReason.unknown, httpCacheItem.policy.responseHeaders());
    }

    if (httpCacheItem.status === 429) {
      const retryAfter = parseInt(`${httpCacheItem.policy.responseHeaders()['retry-after']}`);
      if (!isNaN(retryAfter)) {
        this.rateLimitedCache.set(url.host, undefined, { ttl: retryAfter * 1000 });
      }

      throw new TooManyRequestsError(retryAfter);
    }

    throw new HttpError(httpCacheItem.status, httpCacheItem.statusText, responseHeaders);
  };

  private determineCacheKey(url: URL, init?: CustomRequestInit): string {
    return `${url.href}_${init?.body?.toString()}`;
  }

  private determineCacheTtl(httpCacheItem: HttpCacheItem): number {
    if (httpCacheItem.status === 200) {
      return Math.max(httpCacheItem.policy.timeToLive(), this.MIN_CACHE_TTL);
    }

    return httpCacheItem.policy.timeToLive();
  };

  private cacheGet(key: string): HttpCacheItem | undefined {
    const buffer = this.httpCache.get(key);

    return buffer ? JSON.parse(gunzipSync(buffer).toString()) : undefined;
  }

  private cacheSet(key: string, httpCacheItem: HttpCacheItem) {
    const ttl = this.determineCacheTtl(httpCacheItem);
    if (ttl <= 0) {
      return;
    }

    this.httpCache.set(key, gzipSync(JSON.stringify(httpCacheItem)), { ttl });
  }

  private headersToObject(headers: Headers): Record<string, string> {
    const obj: Record<string, string> = {};

    headers.forEach((value, name) => {
      obj[name] = value;
    });

    return obj;
  };

  private async cachedFetch(ctx: Context, url: URL, init?: CustomRequestInit): Promise<HttpCacheItem> {
    const newInit = this.getInit(ctx, url, init);

    const request: CachePolicy.Request = { url: url.href, method: newInit.method ?? 'GET', headers: {} };

    const cacheKey = this.determineCacheKey(url, init);
    let httpCacheItem = this.cacheGet(cacheKey);
    if (httpCacheItem) {
      this.logger.info(`Cached fetch ${request.method} ${url}`, ctx);
      return this.handleHttpCacheItem(ctx, httpCacheItem, url, init);
    }

    const response = await this.queuedFetch(ctx, url, newInit);
    const body = await response.text();

    const policy = new CachePolicy(request, { status: response.status, headers: this.headersToObject(response.headers) }, { shared: false });

    httpCacheItem = { policy, status: response.status, statusText: response.statusText, body };

    this.cacheSet(cacheKey, httpCacheItem);

    return this.handleHttpCacheItem(ctx, httpCacheItem, url, init);
  };

  protected async fetchWithTimeout(ctx: Context, url: URL, init?: CustomRequestInit): Promise<Response> {
    this.logger.info(`Fetch ${init?.method ?? 'GET'} ${url}`, ctx);

    if (this.rateLimitedCache.has(url.host)) {
      throw new TooManyRequestsError(this.rateLimitedCache.getRemainingTTL(url.host) / 1000);
    }

    if ((this.timeoutsCountCache.get(url.host) ?? 0) >= (init?.timeoutsCountThrow ?? this.DEFAULT_FAILED_REQUEST_COUNT_THROW)) {
      throw new TooManyTimeoutsError();
    }

    let response;
    try {
      response = await fetch(url, { ...init, keepalive: true, signal: AbortSignal.timeout(init?.timeout ?? this.DEFAULT_TIMEOUT) });
    } catch (error) {
      if (error instanceof DOMException && error.name === 'TimeoutError') {
        await this.increaseTimeoutsCount(url);
        throw new TimeoutError();
      }

      throw error;
    }

    await this.decreaseTimeoutsCount(url);

    return response;
  };

  private async increaseTimeoutsCount(url: URL) {
    await this.timeoutsCountMutex.runExclusive(async () => {
      const count = (this.timeoutsCountCache.get(url.host) ?? 0) + 1;
      this.timeoutsCountCache.set(url.host, count, { ttl: this.TIMEOUT_CACHE_TTL });
    });
  }

  private async decreaseTimeoutsCount(url: URL) {
    await this.timeoutsCountMutex.runExclusive(async () => {
      const count = Math.max(0, (this.timeoutsCountCache.get(url.host) ?? 0) - 1);
      this.timeoutsCountCache.set(url.host, count, { ttl: this.TIMEOUT_CACHE_TTL });
    });
  }

  private getSemaphore(url: URL, queueLimit: number, queueTimeout: number): SemaphoreInterface {
    let sem = this.semaphores.get(url.host);

    if (!sem) {
      sem = withTimeout(new Semaphore(queueLimit), queueTimeout, new QueueIsFullError());
      this.semaphores.set(url.host, sem);
    }

    return sem;
  }

  private async queuedFetch(ctx: Context, url: URL, init?: CustomRequestInit): Promise<Response> {
    const queueLimit = init?.queueLimit ?? this.DEFAULT_QUEUE_LIMIT;
    const queueTimeout = init?.queueTimeout ?? this.DEFAULT_QUEUE_TIMEOUT;

    const semaphore = this.getSemaphore(url, queueLimit, queueTimeout);

    const [,release] = await semaphore.acquire();

    try {
      return await this.fetchWithTimeout(ctx, url, init);
    } finally {
      release();
    }
  }
}
