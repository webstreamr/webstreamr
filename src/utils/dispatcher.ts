import { socksDispatcher } from 'fetch-socks';
import { minimatch } from 'minimatch';
import { Dispatcher, ProxyAgent } from 'undici';
import { Context } from '../types';

export const getProxyForUrl = (ctx: Context, url: URL): URL | undefined => {
  const proxyConfig = ctx.config['proxyConfig'] || process.env['PROXY_CONFIG'];

  if (proxyConfig) {
    for (const rule of proxyConfig.split(',')) {
      const [hostPattern, proxy] = rule.split(/:(.+)/);
      if (!hostPattern || !proxy) {
        throw new Error(`Proxy rule "${rule}" is invalid.`);
      }

      if (hostPattern === '*' || minimatch(url.host, hostPattern)) {
        return proxy === 'false' ? undefined : new URL(proxy);
      }
    }
  } else if (process.env['ALL_PROXY']) {
    return new URL(process.env['ALL_PROXY']);
  }

  return undefined;
};

export const createProxyAgent = (proxyUrl: URL): Dispatcher => {
  if (proxyUrl.protocol === 'socks5:') {
    return socksDispatcher({ type: 5, host: proxyUrl.hostname, port: parseInt(proxyUrl.port) });
  }

  return new ProxyAgent({ uri: proxyUrl.href });
};
