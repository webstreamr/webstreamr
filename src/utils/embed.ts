import { unpack } from 'unpacker';

export class CouldNotFindStreamLinkInEmbedError extends Error {}

export const extractUrlFromPacked = (html: string, linkRegExps: RegExp[]): URL => {
  const evalMatch = html.match(/eval\(function\(p,a,c,k,e,d\).*\)\)/);
  if (!evalMatch) {
    throw new Error(`No p.a.c.k.e.d string found`);
  }

  const unpacked = unpack(evalMatch[0]);

  for (const linkRegexp of linkRegExps) {
    const linkMatch = unpacked.match(linkRegexp);
    if (linkMatch && linkMatch[1]) {
      return new URL('https://' + linkMatch[1].replace(/^(https:)?\/\//, ''));
    }
  }

  throw new CouldNotFindStreamLinkInEmbedError(`Could not find a stream link in embed`);
};
