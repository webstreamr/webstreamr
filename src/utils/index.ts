export * from './embed';
export * from './fetch';
export * from './imdb';
export * from './language';
export * from './log';
export * from './manifest';
export * from './promise';
