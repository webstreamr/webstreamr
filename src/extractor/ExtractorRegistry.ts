import TTLCache from '@isaacs/ttlcache';
import winston from 'winston';
import { Extractor } from './types';
import { Context, UrlResult } from '../types';
import { Fetcher } from '../utils';
import { DoodStream } from './DoodStream';
import { Dropload } from './Dropload';
import { SuperVideo } from './SuperVideo';
import { Mixdrop } from './Mixdrop';

export class ExtractorRegistry {
  private readonly logger: winston.Logger;
  private readonly extractors: Extractor[];
  private readonly urlResultCache: TTLCache<string, UrlResult>;

  constructor(logger: winston.Logger, fetcher: Fetcher) {
    this.logger = logger;
    this.extractors = [
      new DoodStream(fetcher),
      new Dropload(fetcher),
      new Mixdrop(fetcher),
      new SuperVideo(fetcher),
    ];
    this.urlResultCache = new TTLCache({ max: 1024 });
  }

  readonly handle = async (ctx: Context, url: URL, countryCode: string): Promise<UrlResult | undefined> => {
    let urlResult = this.urlResultCache.get(url.href);
    if (urlResult) {
      return urlResult;
    }

    const extractor = this.extractors.find(embedExtractor => embedExtractor.supports(url));
    if (undefined === extractor) {
      return undefined;
    }

    this.logger.info(`Extract stream URL using ${extractor.id} extractor from ${url}`);

    try {
      urlResult = await extractor.extract(ctx, url, countryCode);
    } catch (error) {
      this.logger.warn(`${extractor.id} error: ` + error);
      return undefined;
    }

    this.urlResultCache.set(url.href, urlResult, { ttl: extractor.ttl });

    return urlResult;
  };
}
