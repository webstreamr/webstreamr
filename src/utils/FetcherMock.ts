/* istanbul ignore file */
import crypto from 'crypto';
import fs from 'node:fs';
import CachePolicy from 'http-cache-semantics';
import slugify from 'slugify';
import { RequestInit } from 'undici';
import winston from 'winston';
import { NotFoundError } from '../error';
import { Context } from '../types';
import { envGet } from './env';
import { Fetcher, HttpCacheItem } from './Fetcher';

export class FetcherMock extends Fetcher {
  private readonly fixturePath: string;

  public constructor(fixturePath: string) {
    super(winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] }));

    this.fixturePath = fixturePath;
  }

  public override async fetch(ctx: Context, url: URL, init?: RequestInit): Promise<HttpCacheItem> {
    let path: string;

    if (init?.method === 'POST') {
      const body = init.body as string;
      path = `${this.fixturePath}/post-${this.slugifyUrl(url)}-${slugify(body)}`;
    } else if (init?.method === 'HEAD') {
      path = `${this.fixturePath}/head-${this.slugifyUrl(url)}`;
    } else {
      path = `${this.fixturePath}/${this.slugifyUrl(url)}`;
    }

    return this.fetchInternal(path, ctx, url, init);
  };

  public override async text(ctx: Context, url: URL, init?: RequestInit): Promise<string> {
    const path = `${this.fixturePath}/${this.slugifyUrl(url)}`;

    return (await this.fetchInternal(path, ctx, url, init)).body;
  };

  public override async textPost(ctx: Context, url: URL, body: string, init?: RequestInit): Promise<string> {
    const path = `${this.fixturePath}/post-${this.slugifyUrl(url)}-${slugify(body)}`;

    return (await this.fetchInternal(path, ctx, url, { ...init, method: 'POST', body })).body;
  };

  public override async head(ctx: Context, url: URL, init?: RequestInit): Promise<CachePolicy.Headers> {
    const path = `${this.fixturePath}/head-${this.slugifyUrl(url)}`;

    return (await this.fetchInternal(path, ctx, url, { ...init, method: 'HEAD' })).headers;
  };

  private readonly slugifyUrl = (url: URL): string => {
    const slugifiedUrl = slugify(url.href);

    if (slugifiedUrl.length > 249) {
      return slugify(`${url.origin}-${crypto.createHash('md5').update(url.href).digest('hex')}`);
    }

    return slugifiedUrl;
  };

  private readonly fetchInternal = async (path: string, ctx: Context, url: URL, init?: RequestInit): Promise<HttpCacheItem> => {
    const errorPath = `${path}.error`;

    const isHead = init?.method === 'HEAD';

    if (fs.existsSync(errorPath)) {
      const message = fs.readFileSync(errorPath).toString();
      if (message.includes('404: Not Found')) {
        throw new NotFoundError(message);
      }

      throw new Error(message);
    } else if (fs.existsSync(path)) {
      const data = fs.readFileSync(path).toString();

      return {
        body: isHead ? '' : data,
        headers: isHead ? JSON.parse(data) : {},
        status: 200,
        statusText: 'OK',
        ttl: 0,
        url: url.href,
      };
    } else {
      let response;
      try {
        if (envGet('TEST_UPDATE_FIXTURES')) {
          response = await super.fetchWithTimeout(ctx, url, init);
        } else {
          console.error(`No fixture found at "${path}".`);
          process.exit(1);
        }
      } catch (error) {
        fs.writeFileSync(errorPath, `${error}`);
        throw error;
      }

      if (response.status < 200 || response.status > 399) {
        const message = `Fetcher error: ${response.status}: ${response.statusText}`;
        fs.writeFileSync(errorPath, message);
        throw new Error(message);
      }

      let result;
      if (isHead) {
        const headers: Record<string, string> = {};

        response.headers.forEach((value, key) => {
          headers[key] = value;
        });

        result = JSON.stringify(headers);
      } else {
        result = await response.text();
      }

      fs.writeFileSync(path, result);

      return {
        body: isHead ? '' : result,
        headers: isHead ? JSON.parse(result) : {},
        status: 200,
        statusText: 'OK',
        ttl: 0,
        url: url.href,
      };
    }
  };
}
