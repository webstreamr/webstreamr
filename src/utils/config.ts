import { Extractor } from '../extractor';
import { Config } from '../types';

export const getDefaultConfig = (): Config => {
  return { multi: 'on', en: 'on' };
};

export const showErrors = (config: Config): boolean => 'showErrors' in config;

export const showExternalUrls = (config: Config): boolean => 'includeExternalUrls' in config;

export const hasMultiEnabled = (config: Config): boolean => 'multi' in config;

export const disableExtractorConfigKey = (extractor: Extractor): string => `disableExtractor_${extractor.id}`;

export const isExtractorDisabled = (config: Config, extractor: Extractor): boolean => disableExtractorConfigKey(extractor) in config;
