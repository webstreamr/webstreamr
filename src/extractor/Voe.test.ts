import winston from 'winston';
import { createTestContext } from '../test';
import { FetcherMock } from '../utils';
import { ExtractorRegistry } from './ExtractorRegistry';
import { Voe } from './Voe';

const logger = winston.createLogger({ transports: [new winston.transports.Console({ level: 'nope' })] });
const extractorRegistry = new ExtractorRegistry(logger, [new Voe(new FetcherMock(`${__dirname}/__fixtures__/Voe`))]);

const ctx = createTestContext({ mediaFlowProxyUrl: 'https://mediaflow.test.org', mediaFlowProxyPassword: 'test' });

describe('Voe', () => {
  test('jilliandescribecompany', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://jilliandescribecompany.com/e/ea21l02gcygw'))).toMatchSnapshot();
  });

  test('premium only without resolution', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://jilliandescribecompany.com/qqfyi04w52mj'))).toMatchSnapshot();
  });

  test('encoding error', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://voe.sx/e/c2yxvit4f6bz'))).toMatchSnapshot();
  });

  test('embed only urls which otherwise lead to 404', async () => {
    expect(await extractorRegistry.handle(ctx, new URL('https://mikaylaarealike.com/e/gqlhm9hbobwu'))).toMatchSnapshot();
  });
});
