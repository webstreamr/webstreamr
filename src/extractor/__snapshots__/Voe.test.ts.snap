// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Voe embed only urls which otherwise lead to 404 1`] = `
[
  {
    "format": "hls",
    "label": "VOE (via MediaFlow Proxy)",
    "meta": {
      "countryCodes": [],
      "height": 800,
      "title": "Weapons.2025.German.DL.1080p.WEB.H264-ZeroTwo_encoded.mp4",
    },
    "sourceId": "voe_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fjilliandescribecompany.com%2Fqqfyi04w52mj&d=https%3A%2F%2Fcdn-thhhlmzuxvpehc9r.orbitcache.com%2Fengine%2Fhls2-c%2F01%2F11526%2Fqqfyi04w52mj_%2Cl%2C.urlset%2Fmaster.m3u8%3Ft%3Dg2zuS_OAHTtjrVe0bHp2xYcX7PpDxg8ebRIHTmF3H1c%26s%3D1759861731%26e%3D14400%26f%3D57634053%26node%3DUOx5276IumOH4BADikx4rlHC%2BibdAWZ%2Bmo07sx3WfaU%3D%26i%3D104.28%26sp%3D2500%26asn%3D13335%26q%3Dl%26rq%3DFxkyWdISXylA0UaLSPNcXWkDrksjs0pkAzy0GvIS",
  },
]
`;

exports[`Voe encoding error 1`] = `[]`;

exports[`Voe jilliandescribecompany 1`] = `
[
  {
    "format": "hls",
    "label": "VOE (via MediaFlow Proxy)",
    "meta": {
      "bytes": 1202590842,
      "countryCodes": [],
      "height": 720,
      "title": "Superman.2025.2160p.WEB-DL.DV.HDR10Plus.X265.DDP5.1.Atmos-YIN.mp4",
    },
    "sourceId": "voe_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fjilliandescribecompany.com%2Fe%2Fea21l02gcygw&d=https%3A%2F%2Fcdn-kuc3ehpz4wnczfyz.orbitcache.com%2Fengine%2Fhls2-c%2F01%2F15242%2Fea21l02gcygw_%2Cn%2C.urlset%2Fmaster.m3u8%3Ft%3D3iMAKoJqij6nC4HDamYxtDB75a_-AlG_wD1demRBo0I%26s%3D1759754183%26e%3D14400%26f%3D76213694%26node%3DFPYruvgfgORaUYuYfpByMgjFbC8X%2FHWKW1utbUQQRyQ%3D%26i%3D104.28%26sp%3D2500%26asn%3D13335%26q%3Dn%26rq%3DzKNpONzQJHzTeXTzr0V2M9V4DUnzwX9C8hTNxrUW",
  },
]
`;

exports[`Voe premium only without resolution 1`] = `
[
  {
    "format": "hls",
    "label": "VOE (via MediaFlow Proxy)",
    "meta": {
      "bytes": 363216240,
      "countryCodes": [],
      "height": 800,
      "title": "268836--21ba9fb2-97f9-4566-b5eb-3d9cef7674cf--zrxn--2352322-voesx.mp4",
    },
    "sourceId": "voe_",
    "ttl": 900000,
    "url": "https://mediaflow.test.org/proxy/hls/manifest.m3u8?api_password=test&h_user-agent=Mozilla%2F5.0+%28Windows+NT+10.0%3B+Win64%3B+x64%29+AppleWebKit%2F537.36+%28KHTML%2C+like+Gecko%29+Chrome%2F136.0.0.0+Safari%2F537.36&h_referer=https%3A%2F%2Fjilliandescribecompany.com%2Fqqfyi04w52mj&d=https%3A%2F%2Fcdn-thhhlmzuxvpehc9r.orbitcache.com%2Fengine%2Fhls2-c%2F01%2F11526%2Fqqfyi04w52mj_%2Cl%2C.urlset%2Fmaster.m3u8%3Ft%3Dg2zuS_OAHTtjrVe0bHp2xYcX7PpDxg8ebRIHTmF3H1c%26s%3D1759861731%26e%3D14400%26f%3D57634053%26node%3DUOx5276IumOH4BADikx4rlHC%2BibdAWZ%2Bmo07sx3WfaU%3D%26i%3D104.28%26sp%3D2500%26asn%3D13335%26q%3Dl%26rq%3DFxkyWdISXylA0UaLSPNcXWkDrksjs0pkAzy0GvIS",
  },
]
`;
